### PR TITLE
Llama index doc

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12692?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12692/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12692
```

</p>
</details>

### What changes are proposed in this pull request?

Add the main flavor page for LlamaIndex.

**Note**: Tracing page will be updated once the prep PR https://github.com/mlflow/mlflow/pull/12691 is merged

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/llms/llama-index/index.rst [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+Why use LlamaIndex with MLflow?
+-------------------------------
+
+The integration of the LlamaIndex library with MLflow provides a seamless experience for managing and deploying LlamaIndex engines. The following are some of the key benefits of using LlamaIndex with MLflow:
+
+* `MLflow Tracking <../../tracking.html>`_ allows you to track your index in MLflow and manage many moving parts in your LlamaIndex project, such as prompts, LLMs, retrievers, tools, global configurations, and more.
+
+* `MLflow Model <../../models.html>`_ packages your LlamaIndex engine with all its dependency versions, input and output interfaces, and other essential metadata. This allows you to deploy your LlamaIndex engine with ease, knowing that the environment is consistent across different stages of the ML lifecycle.
+
+* `MLflow Evaluate <../llm-evaluate/index.html>`_ provides native capabilities within MLflow to evaluate language models. This capability facilitates the efficient assessment of inference results from your LlamaIndex engine, ensuring robust performance analytics and facilitating quick iterations.
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:32]:
```suggestion
* `MLflow Evaluate <../llm-evaluate/index.html>`_ provides native capabilities within MLflow to evaluate GenAI applications. This capability facilitates the efficient assessment of inference results from your LlamaIndex engine, ensuring robust performance analytics and facilitating quick iterations.
```

File: docs/source/llms/llama-index/index.rst [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+* `MLflow Tracing <../tracing/index.html>`_ is a powerful observability tool for monitoring and debugging what happens inside the LlamaIndex models, helping you identify potential bottlenecks or issues quickly. With its powerful automatic logging capability, you can instrument your LlamaIndex application without any code change but just by running a single command.
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:34]:
```suggestion
* `MLflow Tracing <../tracing/index.html>`_ is a powerful observability tool for monitoring and debugging what happens inside the LlamaIndex models, helping you identify potential bottlenecks or issues quickly. With its powerful automatic logging capability, you can instrument your LlamaIndex application without needing to add any code apart from running a single command.
```

File: docs/source/llms/llama-index/index.rst [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+TBA
+
+
+Concepts
+--------
+
+``Index``
+^^^^^^^^^
+
+The ``Index`` object is the core foundation in LlamaIndex, which is a collection of documents that are indexed for fast information retrieval. Such as Retrieval-Augmented Generation (RAG) and Agents use cases. The ``Index`` object can be logged to MLflow experiments and loaded back as an inference engine.
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:50]:
```suggestion
The ``Index`` object is the core foundation in LlamaIndex. This object is a collection of documents that are indexed for fast information retrieval, providing capabilities for applications such as Retrieval-Augmented Generation (RAG) and Agents. The ``Index`` object can be logged directly to an MLflow run and loaded back for use as an inference engine.
```

File: docs/source/llms/llama-index/index.rst [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+.. toctree::
+    :maxdepth: 2
+
+Saving and Loading Index in MLflow Experiment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Creating an Index
+~~~~~~~~~~~~~~~~~
+
+The ``index`` object is a centerpiece of the LlamaIndex and MLflow integration. With LlamaIndex, you can create an index from a collection of documents. The following code creates a sample index from Paul Graham's essay data available in the LlamaIndex repository.
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:72]:
```suggestion
The ``index`` object is the centerpiece of the LlamaIndex and MLflow integration. With LlamaIndex, you can create an index from a collection of documents. The following code creates a sample index from Paul Graham's essay data available within the LlamaIndex repository.
```

File: docs/source/llms/llama-index/index.rst [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+One key step here is to specify the ``engine_type`` parameter. The choice of engine type does not affect the index itself,
+but dictates the interface of how you query the index when you load it back for inference.
+
+1. QueryEngine (``engine_type="query"``) is designed for a simple query-response system that takes a single query string and returns a response.
+2. ChatEngine (``engine_type="chat"``) is designed for a conversational agent that keeps track of the conversation history and answer tp the user query based on the context.
+3. Retriever (``engine_type="retriever"``) is a lower-level component that returns the top-k relevant documents matching the query.
+
+
+The following code is an example of logging the index to MLflow with chat engine type.
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:99]:
```suggestion
The following code is an example of logging an index to MLflow with the ``chat`` engine type.
```

File: docs/source/llms/llama-index/index.rst [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    # >> The first program the author wrote was on the IBM 1401 ...
+
+    # The chat engine keeps track of the conversation history
+    response = model.predict("How did the author feel about it?")
+    print(response)
+    # >> The author felt puzzled by the first program ...
+
+
+.. tip::
+
```
Comment (BenWilson2) [inline docs/source/llms/llama-index/index.rst:147]: We might want to pre-emptively add the next logical question to an FAQ section for users. 
"Ok, I saved my LlamaIndex 'model' with a 'chat engine', but now I want to use it as a 'query engine' deployed to an endpoint" - we could walk them through the process of extracting the index and saving with the alternative engine type and showing how the response differs in tracing to fully illustrate the versatility of the design choices for this flavor (the index is huge and expensive to recreate, the engine is easy to swap out - let's just connect the dots for them)

**Review by BenWilson2 — APPROVED** [review]
Minor nits. LGTM once an expanded guidance is added for 'swapping' engines via native loader and re-logging

File: docs/source/llms/llama-index/index.rst (hunk 6 above)
Comment (B-Step62) [reply docs/source/llms/llama-index/index.rst:147]: Yea I initially added the FAQ section heading and couldn't come up with a good question candidates😅 The engine switching question seems to be a great one to start!

File: docs/source/llms/llama-index/index.rst [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+.. note::
+
+    The tracing supports async prediction and streaming response, however, it does not
+    support the combination of async and streaming, such as the ``astream_chat`` method.
+
+
+FAQ
+---
+
+I have an index logged with ``query`` engine type. Can I load it back a ``chat`` engine?
```
Comment (B-Step62) [inline docs/source/llms/llama-index/index.rst:192]:
```suggestion
I have an index logged with ``query`` engine type. Can I load it back as a ``chat`` engine?
```

typo

File: docs/source/llms/llama-index/index.rst [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (B-Step62) [inline docs/source/llms/llama-index/index.rst:193]:
```suggestion
I have an index logged with ``query`` engine type. Can I load it back as a ``chat`` engine?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/source/llms/llama-index/index.rst [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+^^^^^^^^^
+
+The ``Index`` object is the core foundation in LlamaIndex. This object is a collection of documents that are indexed for fast information retrieval, providing capabilities for applications such as Retrieval-Augmented Generation (RAG) and Agents. The ``Index`` object can be logged directly to an MLflow run and loaded back for use as an inference engine.
+
+
+``Engine``
+^^^^^^^^^^
+
+The ``Engine`` is a generic interface built on top of the ``Index`` object, which provides a set of APIs to interact with the index. LlamaIndex provides two types of engines: ``QueryEngine`` and ``ChatEngine``. The ``QueryEngine`` simply takes a single
+query and returns a response based on the index. The ``ChatEngine`` is designed for conversational agents, which keeps track of the conversation history as well.
```
Comment (michael-berk) [inline docs/source/llms/llama-index/index.rst:57]: Probably worth mentioning retrievers in here as well even though they're not engines.

File: docs/source/llms/llama-index/index.rst [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+.. toctree::
+    :maxdepth: 2
+
+Saving and Loading Index in MLflow Experiment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Creating an Index
+~~~~~~~~~~~~~~~~~
+
+The ``index`` object is the centerpiece of the LlamaIndex and MLflow integration. With LlamaIndex, you can create an index from a collection of documents. The following code creates a sample index from Paul Graham's essay data available within the LlamaIndex repository.
```
Comment (michael-berk) [inline docs/source/llms/llama-index/index.rst:72]:
```suggestion
The ``index`` object is the centerpiece of the LlamaIndex and MLflow integration. With LlamaIndex, you can create an index from a collection of documents or external vector stores. The following code creates a sample index from Paul Graham's essay data available within the LlamaIndex repository.
```

File: docs/source/llms/llama-index/index.rst [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+    documents = SimpleDirectoryReader("data").load_data()
+    index = VectorStoreIndex.from_documents(documents)
+
+Logging the Index to MLflow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can log the ``index`` object to the MLflow experiment using the :py:func:`mlflow.llama_index.log_model` function.
+
+One key step here is to specify the ``engine_type`` parameter. The choice of engine type does not affect the index itself,
```
Comment (michael-berk) [inline docs/source/llms/llama-index/index.rst:91]:
```suggestion
One key step here is to specify the ``engine_type`` parameter, which can take the values: "chat", "query", and "retriever". The choice of engine type does not affect the index itself,
```

File: docs/source/llms/llama-index/index.rst [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+but dictates the interface of how you query the index when you load it back for inference.
+
+1. QueryEngine (``engine_type="query"``) is designed for a simple query-response system that takes a single query string and returns a response.
+2. ChatEngine (``engine_type="chat"``) is designed for a conversational agent that keeps track of the conversation history and answer tp the user query based on the context.
```
Comment (michael-berk) [inline docs/source/llms/llama-index/index.rst:95]:
```suggestion
2. ChatEngine (``engine_type="chat"``) is designed for a conversational agent that keeps track of the conversation history and responds to a user message.
```

File: docs/source/llms/llama-index/index.rst (hunk 6 above)
Comment (michael-berk) [reply docs/source/llms/llama-index/index.rst:147]: Another good one is stuff around settings e.g. how should I change the LLM/embedding model of a loaded index. Not sure how much llama index education we want to do though 🙂

File: docs/source/llms/llama-index/index.rst [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+
+    # Load the index back and re-log it with the chat engine type
+    index = mlflow.llama_index.load_model(model_info.model_uri)
+    with mlflow.start_run():
+        model_info = mlflow.llama_index.log_model(
+            index,
+            artifact_path="index-chat",
+            # Specify the chat engine type this time
+            engine_type="chat",
+        )
```
Comment (michael-berk) [inline docs/source/llms/llama-index/index.rst:220]:
```suggestion
        )
        
Also note that an index loaded to memory is a llama_index object, so you can leverage their standard inference APIs, specifically:
* ``index.as_chat_engine().chat("hi")``
* ``index.as_query_engine().query("hi")``
* ``index.as_retriever_engine().retrieve("hi")``
```

File: mlflow/llama_index/__init__.py [hunk 14]
```diff
@@ -363,6 +365,8 @@ def autolog(
         silent: If ``True``, suppress all event logs and warnings from MLflow during LlamaIndex
             autologging. If ``False``, show all events and warnings.
     """
+    from mlflow.llama_index.tracer import remove_llama_index_tracer, set_llama_index_tracer
```
Comment (michael-berk) [inline mlflow/llama_index/__init__.py:368]: (for learning purposes) Why did you move the imports into the functions? Readability?

**Review by michael-berk — APPROVED** [review]

File: mlflow/llama_index/__init__.py (hunk 14 above)
Comment (B-Step62) [reply mlflow/llama_index/__init__.py:368]: Great question! This is indeed required for doc building. The python API doc is auto-generated from docstring, but in CI we don't install LlamaIndex (and many other ML libraries for other flavors), so if we do `import llama_index` at the top level it will fail with import error🙂
